File: ocaml/apiroximate.ml
open Complex

let approximation count =
    let random () = {re = Random.float 1.; im = Random.float 1.} in
    let rec randoms = function
        | 0 -> []
        | n -> random () :: randoms (n - 1) in
    let in_circle = List.length (List.filter (fun p -> norm p <= 1.) (randoms count)) in
    4.0 *. float in_circle /. float count;;

for count = 1 to 100000 do
    Printf.printf "%6d %f" count (approximation count);
    print_newline()
done

File: clojure/a-pi-roximate.clj
(defrecord Point [x y])

(defn dist [p]
  (let [x (:x p) y (:y p)]
    (Math/sqrt (+ (* x x) (* y y)))))

(defn approximation [c]
  (letfn
    [(random-point [] (Point. (rand) (rand)))
     (random-points [c]
       (if (zero? c) []
         (cons (random-point) (random-points (dec c)))))]
    (let [in-circle (count (filter (fn [p] (<= (dist p) 1)) (random-points c)))]
      (* 4 (/ in-circle c)))))

(for [c (range 100000)]
  (printf "%6d %f\n" c (approximation c)))
